File: examples/testlib/art/cartoon/stickman/zoomWalkMan.js
const DEFAULT_POSITION = [0, 10];

function zoomWalkMan(optionsInput = {}) {
  const options = ensureObject(optionsInput);
  const anchorBase = toPoint(options.initialPosition, DEFAULT_POSITION);
  const measurementsInput = ensureObject(options.initialMeasurements);
  const depthDelta = toNumber(options.depthDelta, -12); // total vertical change (world Y) for moving foot
  const progress = clamp01(toNumber(options.progress, 0));
  const direction = normalizeDirection(options.direction, "front");
  const zoomTarget = Math.max(0, toNumber(options.zoom, 1));
  const stepper = steperManZoom;

  const defaultOffsets = resolveDefaultLegOffsets(measurementsInput);
  const defaultStride = resolveStrideLength(measurementsInput, defaultOffsets);
  const strideDirection = depthDelta >= 0 ? 1 : -1;
  const spacing = Math.max(1e-6, defaultStride);
  const overreach = Math.max(spacing * 0.25, 0.75);
  const passDistance = spacing + overreach;
  const stepCount = Math.max(1, Math.ceil(Math.abs(depthDelta) / passDistance));
  const totalProgress = progress * stepCount;
  const activeStepIndex = Math.min(stepCount - 1, Math.floor(totalProgress));
  const activeStepPhase = totalProgress - activeStepIndex;

  let currentAnchor = anchorBase;
  let currentMeasurements = mergeFacing(measurementsInput, direction);
  let remainingDepth = Math.abs(depthDelta);
  let leftFoot = addPoints(currentAnchor, resolveLegOffset(currentMeasurements?.legs?.left, defaultOffsets.left));
  let rightFoot = addPoints(currentAnchor, resolveLegOffset(currentMeasurements?.legs?.right, defaultOffsets.right));

  // pick moving side: move the deeper foot first when zooming in (negative delta), shallower when zooming out
  const movingSideInitial = strideDirection >= 0
    ? (leftFoot[1] >= rightFoot[1] ? "left" : "right")
    : (leftFoot[1] <= rightFoot[1] ? "left" : "right");
  const fixedSideInitial = movingSideInitial === "left" ? "right" : "left";

  for (let i = 0; i < stepCount; i += 1) {
    const stepFixedSide = i % 2 === 0 ? fixedSideInitial : movingSideInitial;
    const stepMovingSide = stepFixedSide === "left" ? "right" : "left";
    const isActiveStep = i === activeStepIndex;
    const stepProgress = i < activeStepIndex ? 1 : isActiveStep ? activeStepPhase : 0;
    const remainingSteps = Math.max(1, stepCount - i);

    const fixedY = stepFixedSide === "left" ? leftFoot[1] : rightFoot[1];
    let strideMagnitude;
    if (remainingSteps === 1) {
      strideMagnitude = remainingDepth;
    } else {
      const reserve = (remainingSteps - 1) * passDistance;
      const allowed = Math.max(passDistance, remainingDepth - reserve);
      strideMagnitude = Math.min(allowed, remainingDepth);
    }
    const targetY = fixedY + strideMagnitude * strideDirection;

    const movingStart = stepMovingSide === "left" ? leftFoot : rightFoot;
    const stepFixedPoint = stepFixedSide === "left" ? leftFoot : rightFoot;

    const measurementsWithOffsets = applyLegOffsets(currentMeasurements, currentAnchor, leftFoot, rightFoot, defaultOffsets);

    const zoomPhase = (i + stepProgress) / stepCount;
    const currentZoom = 1 + (zoomTarget - 1) * zoomPhase;

    const result = stepper({
      position: currentAnchor,
      measurements: measurementsWithOffsets,
      movingSide: stepMovingSide,
      movingFootTargetY: targetY,
      zoom: currentZoom,
      progress: stepProgress
    });

    const resolvedMeasurements = ensureObject(
      result?.measurements || result?.sequenceState?.measurements
    ) || measurementsWithOffsets;
    const resolvedPosition = toPoint(
      result?.position || result?.sequenceState?.position,
      currentAnchor
    );

    currentMeasurements = resolvedMeasurements;
    currentAnchor = resolvedPosition;

    leftFoot = addPoints(currentAnchor, resolveLegOffset(currentMeasurements?.legs?.left, defaultOffsets.left));
    rightFoot = addPoints(currentAnchor, resolveLegOffset(currentMeasurements?.legs?.right, defaultOffsets.right));
    remainingDepth = Math.max(0, remainingDepth - strideMagnitude);

    if (isActiveStep) {
      break;
    }
  }

  return {
    measurements: cloneValue(currentMeasurements),
    position: cloneValue(currentAnchor)
  };
}

function resolveLegOffset(measurement, fallback) {
  return toPoint(measurement?.effectorCoordinate, fallback);
}

function resolveDefaultLegOffsets(measurements, fallback = { left: [-2, -11], right: [2, -11] }) {
  const legs = ensureObject(measurements?.legs);
  return {
    left: toPoint(legs.left?.effectorCoordinate, fallback.left),
    right: toPoint(legs.right?.effectorCoordinate, fallback.right)
  };
}

function resolveStrideLength(measurements, defaults) {
  const legs = ensureObject(measurements?.legs);
  const left = ensureObject(legs.left);
  const right = ensureObject(legs.right);
  const leftLength = Math.max(0, toNumber(left.upperLength, 0) + toNumber(left.lowerLength, 0));
  const rightLength = Math.max(0, toNumber(right.upperLength, 0) + toNumber(right.lowerLength, 0));
  const avgLength = averageNumbers([leftLength, rightLength]);
  if (avgLength > 0) {
    return Math.max(1, avgLength * 0.35);
  }
  const defaultSpacing = Math.abs(defaults.right[1] - defaults.left[1]) || 1;
  return defaultSpacing;
}

function applyLegOffsets(measurements, anchor, leftFoot, rightFoot, defaults) {
  const base = ensureObject(measurements);
  const legs = ensureObject(base.legs);
  const leftBase = ensureObject(legs.left);
  const rightBase = ensureObject(legs.right);
  return {
    ...base,
    legs: {
      left: {
        ...leftBase,
        effectorCoordinate: subtractPoints(leftFoot, anchor) || defaults.left
      },
      right: {
        ...rightBase,
        effectorCoordinate: subtractPoints(rightFoot, anchor) || defaults.right
      }
    }
  };
}

function mergeFacing(measurements, direction) {
  const base = ensureObject(measurements);
  const torso = ensureObject(base.torso);
  const head = ensureObject(base.head);
  const resolvedDirection = normalizeDirection(torso.direction || head.direction, direction);
  return {
    ...base,
    torso: { ...torso, direction: resolvedDirection },
    head: { ...head, direction: resolvedDirection }
  };
}

function normalizeDirection(value, fallback = "front") {
  const text = typeof value === "string" ? value.trim().toLowerCase() : "";
  if (text === "front" || text === "back" || text === "left" || text === "right") {
    return text;
  }
  return fallback;
}

function clamp01(value) {
  const num = toNumber(value, 0);
  if (!Number.isFinite(num)) return 0;
  if (num < 0) return 0;
  if (num > 1) return 1;
  return num;
}

function toNumber(value, fallback = 0) {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function toPoint(value, fallback = [0, 0]) {
  if (Array.isArray(value) && value.length >= 2) {
    return [toNumber(value[0], fallback[0]), toNumber(value[1], fallback[1])];
  }
  if (value && typeof value === "object") {
    if ("x" in value || "y" in value) {
      return [toNumber(value.x, fallback[0]), toNumber(value.y, fallback[1])];
    }
    if ("left" in value || "top" in value) {
      return [toNumber(value.left, fallback[0]), toNumber(value.top, fallback[1])];
    }
  }
  return Array.isArray(fallback) ? [...fallback] : [0, 0];
}

function addPoints(a, b) {
  return [toNumber(a[0], 0) + toNumber(b[0], 0), toNumber(a[1], 0) + toNumber(b[1], 0)];
}

function subtractPoints(a, b) {
  return [toNumber(a[0], 0) - toNumber(b[0], 0), toNumber(a[1], 0) - toNumber(b[1], 0)];
}

function averageNumbers(values) {
  let sum = 0;
  let count = 0;
  for (const value of values) {
    if (typeof value === "number" && Number.isFinite(value)) {
      sum += value;
      count += 1;
    }
  }
  return count > 0 ? sum / count : 0;
}

function ensureObject(value) {
  return value && typeof value === "object" ? value : {};
}

function cloneValue(value) {
  if (Array.isArray(value)) {
    return value.map(cloneValue);
  }
  if (value && typeof value === "object") {
    const clone = {};
    for (const key of Object.keys(value)) {
      clone[key] = cloneValue(value[key]);
    }
    return clone;
  }
  return value;
}



if (typeof module !== "undefined") {
  module.exports = zoomWalkMan;
}

return zoomWalkMan;

File: examples/testlib/art/cartoon/stickman/zoomWalkMan.fs
(optionsInput)=>
{
  helpers:cartoon.helpers;
  defaults:{ position:[0,10]; depthDelta:-12; zoom:1; direction:"front" };
  fallbackOffsets:{ left:[-2,-11]; right:[2,-11] };

  num:(value, fallback)=> if value = null then fallback else value;

  options:optionsInput ?? {};
  anchorBase:if options.initialPosition = null then defaults.position else options.initialPosition;
  measurementsInput:options.initialMeasurements ?? {};
  depthDelta:num(options.depthDelta, defaults.depthDelta);
  progress:clamp01(num(options.progress, 0));
  direction:normalizeDirection(options.direction, defaults.direction);
  zoomTarget:math.max(0, num(options.zoom, defaults.zoom));
  stepper:if options.stepper != null then options.stepper else steperManZoom;

  defaultOffsets:resolveDefaultLegOffsets(measurementsInput, fallbackOffsets);
  strideDirection:if depthDelta >= 0 then 1 else -1;
  strideLength:resolveStrideLength(measurementsInput, defaultOffsets);
  spacing:math.max(0.000001, strideLength);
  overreach:math.max(spacing * 0.25, 0.75);
  passDistance:spacing + overreach;
  stepCount:math.max(1, math.ceil(math.abs(depthDelta) / passDistance));
  totalProgress:progress * stepCount;
  activeStepIndex:math.min(stepCount - 1, math.floor(totalProgress));
  activeStepPhase:totalProgress - activeStepIndex;

  mergedMeasurements:mergeFacing(measurementsInput, direction);
  initialLeftFoot:helpers.addOffset(anchorBase, resolveLegOffset(mergedMeasurements.legs?.left, defaultOffsets.left));
  initialRightFoot:helpers.addOffset(anchorBase, resolveLegOffset(mergedMeasurements.legs?.right, defaultOffsets.right));
  remainingDistance:math.abs(depthDelta);

  movingSideInitial:if strideDirection >= 0
    then if initialLeftFoot[1] >= initialRightFoot[1] then "left" else "right"
    else if initialLeftFoot[1] <= initialRightFoot[1] then "left" else "right";
  fixedSideInitial:if movingSideInitial = "left" then "right" else "left";

  initialState:{
    anchor:anchorBase;
    measurements:mergedMeasurements;
    leftFoot:initialLeftFoot;
    rightFoot:initialRightFoot;
    remaining:remainingDistance;
  };

  finalState:simulateWalk(0, initialState);

  eval {
    measurements:{} + finalState.measurements;
    position:if finalState.anchor = null then defaults.position else finalState.anchor;
  };

  simulateWalk:(index, state)=> {
    eval if index >= stepCount then state else {
      stepFixedSide:if index % 2 = 0 then fixedSideInitial else movingSideInitial;
      stepMovingSide:if stepFixedSide = "left" then "right" else "left";
      isActive:index = activeStepIndex;
      stepProgress:if index < activeStepIndex then 1 else if isActive then activeStepPhase else 0;
      remainingSteps:math.max(1, stepCount - index);
      fixedY:if stepFixedSide = "left" then state.leftFoot[1] else state.rightFoot[1];

      strideMagnitude:if remainingSteps = 1 then state.remaining else {
        reserve:(remainingSteps - 1) * passDistance;
        allowed:math.max(passDistance, state.remaining - reserve);
        eval math.min(allowed, state.remaining);
      };

      targetY:fixedY + strideMagnitude * strideDirection;
      measurementsWithOffsets:applyLegOffsets(state.measurements, state.anchor, state.leftFoot, state.rightFoot, defaultOffsets);
      zoomPhase:(index + stepProgress) / stepCount;
      currentZoom:1 + (zoomTarget - 1) * zoomPhase;

      stepResult:stepper({
        position:state.anchor;
        measurements:measurementsWithOffsets;
        movingSide:stepMovingSide;
        movingFootTargetY:targetY;
        zoom:currentZoom;
        progress:stepProgress;
      });

      resolvedMeasurements:if stepResult?.measurements != null then stepResult.measurements else if stepResult?.sequenceState?.measurements != null then stepResult.sequenceState.measurements else measurementsWithOffsets;
      resolvedAnchor:if stepResult?.position != null then stepResult.position else if stepResult?.sequenceState?.position != null then stepResult.sequenceState.position else state.anchor;

      updatedLeftFoot:helpers.addOffset(resolvedAnchor, resolveLegOffset(resolvedMeasurements.legs?.left, defaultOffsets.left));
      updatedRightFoot:helpers.addOffset(resolvedAnchor, resolveLegOffset(resolvedMeasurements.legs?.right, defaultOffsets.right));
      updatedRemaining:math.max(0, state.remaining - strideMagnitude);

      updatedState:{
        anchor:resolvedAnchor;
        measurements:resolvedMeasurements;
        leftFoot:updatedLeftFoot;
        rightFoot:updatedRightFoot;
        remaining:updatedRemaining;
      };

      eval if isActive then updatedState else simulateWalk(index + 1, updatedState);
    };
  };

  resolveLegOffset:(measurement, fallback)=> {
    legInput:measurement ?? {};
    eval if legInput.effectorCoordinate = null then fallback else legInput.effectorCoordinate;
  };

  resolveDefaultLegOffsets:(measurements, fallback)=> {
    legs:(measurements ?? {}).legs ?? {};
    eval {
      left:if legs.left?.effectorCoordinate = null then fallback.left else legs.left.effectorCoordinate;
      right:if legs.right?.effectorCoordinate = null then fallback.right else legs.right.effectorCoordinate;
    };
  };

  resolveStrideLength:(measurements, defaults)=> {
    legs:(measurements ?? {}).legs ?? {};
    left:legs.left ?? {};
    right:legs.right ?? {};
    leftLength:math.max(0, num(left.upperLength, 0) + num(left.lowerLength, 0));
    rightLength:math.max(0, num(right.upperLength, 0) + num(right.lowerLength, 0));
    avgLength:(leftLength + rightLength) / (if leftLength > 0 and rightLength > 0 then 2 else if leftLength > 0 or rightLength > 0 then 1 else 0);
    eval if avgLength > 0 then math.max(1, avgLength * 0.35) else {
      fallbackSpan:math.max(
        math.abs(defaults.left[1]),
        math.abs(defaults.right[1]),
        math.abs(defaults.right[1] - defaults.left[1])
      );
      eval math.max(1, fallbackSpan * 0.35);
    };
  };

  applyLegOffsets:(measurements, anchor, leftFoot, rightFoot, defaults)=> {
    base:measurements ?? {};
    legs:base.legs ?? {};
    anchorPoint:if anchor = null then [0,0] else anchor;
    leftPoint:if leftFoot = null then helpers.addOffset(anchorPoint, defaults.left) else leftFoot;
    rightPoint:if rightFoot = null then helpers.addOffset(anchorPoint, defaults.right) else rightFoot;
    leftOffset:subtractPoints(leftPoint, anchorPoint);
    rightOffset:subtractPoints(rightPoint, anchorPoint);
    eval base + {
      legs:{
        left:(legs.left ?? {}) + { effectorCoordinate:leftOffset };
        right:(legs.right ?? {}) + { effectorCoordinate:rightOffset };
      };
    };
  };

  mergeFacing:(measurements, facing)=> {
    base:measurements ?? {};
    torso:base.torso ?? {};
    head:base.head ?? {};
    resolvedDirection:normalizeDirection(if torso.direction != null then torso.direction else head.direction, facing);
    eval base + { torso:{ direction:resolvedDirection }; head:{ direction:resolvedDirection } };
  };

  normalizeDirection:(value, fallback)=> {
    defaultDir:fallback ?? "front";
    textValue:if value = null then "" else text.lower(format(value));
    eval if textValue = "left" then "left"
    else if textValue = "right" then "right"
    else if textValue = "front" then "front"
    else if textValue = "back" then "back"
    else defaultDir;
  };

  clamp01:(value)=> {
    v:num(value, 0);
    eval if v < 0 then 0 else if v > 1 then 1 else v;
  };

  subtractPoints:(a, b)=> {
    aSafe:if a = null then [0,0] else a;
    bSafe:if b = null then [0,0] else b;
    eval [
      num(aSafe[0], 0) - num(bSafe[0], 0),
      num(aSafe[1], 0) - num(bSafe[1], 0)
    ];
  };
}
